segmentation_labeling_app/transforms/transform_pipeline.py

```diff
@@ -0,0 +1,159 @@
+    pass
+
+
+class TransformPipelineSchema(argschema.ArgSchema):
+    segmentation_run_id = argschema.fields.Int(
+        required=True,
+        description=("which segmentation_run_id to transform. "
+                     "If not provided, will attempt to query using args: "
+                     "ophys_experiment_id, ophys_segmentation_commit_hash."))
+    ophys_experiment_id = argschema.fields.Int(
+        required=False,
+        description=("used as a query filter if "
+                     "segmentation_run_id is not provided."))
+    ophys_segmentation_commit_hash = argschema.fields.Str(
+        required=False,
+        description=("used as a query filter if "
+                     "segmentation_run_id is not provided."))
+    artifact_basedir = argschema.fields.OutputDir(
+        required=True,
+        description=("artifacts will be written to "
+                     "artifact_basedir/segmentation_run_id."))
+    cropped_shape = argschema.fields.List(
+        argschema.fields.Int,
+        required=True,
+        default=[128, 128],
+        description="[h, w] of bounding box around ROI images/videos")
+    quantile = argschema.fields.Float(
+        required=True,
+        default=0.1,
+        description=("quantile threshold for outlining an ROI. "))
+
+    @mm.pre_load
+    def set_segmentation_run_id(self, data, **kwargs):
+        """if segmentation_run_id not provided, query the database
+        using the other provided args
+        """
+        if "segmentation_run_id" not in data:
+            if not all(i in data for i in [
+                    "ophys_experiment_id",
+                    "ophys_segmentation_commit_hash"]):
+                raise TransformPipelineException(
+                        "if omitting arg segmentation_run_id, must "
+                        "include args ophys_experiment_id and "
+                        "ophys_segmentation_commit_hash")
+            data["segmentation_run_id"] = 12345
+            label_vars = query_utils.get_labeling_env_vars()
+            query_string = (
+                "SELECT id FROM segmentation_runs WHERE "
+                f"ophys_experiment_id={data['ophys_experiment_id']} AND "
+                "ophys_segmentation_commit_hash="
+                f"'{data['ophys_segmentation_commit_hash']}'")
+            entries = query_utils.query(
+                    query_string,
+                    label_vars.user,
+                    label_vars.host,
+                    label_vars.database,
+                    label_vars.password,
+                    label_vars.port)
+            if len(entries) != 1:
+                raise TransformPipelineException(
+                    f"{query_string} did not return exactly 1 result")
+            data['segmentation_run_id'] = entries[0]['id']
+        return data
+
+
+class TransformPipeline(argschema.ArgSchemaParser):
+    default_schema = TransformPipelineSchema
+
+    def run(self):
```

Comment:
Could we separate the logic of the run function into a non class method like in suite2p_job. I think it makes the unit testing more concise and easier to read when this is the case.

Reply:
I suppose we could, but this is a pipeline that runs a bunch of unit-tested functions. Running test code on this class is pretty much an integration test. Given that, I don't see the advantage to your suggestion. Does what I am saying make sense, or could you explain more?

Reply:
I think writing that integration test is cleaner when they are separated is what I'm getting at. I think it's easier to understand the test code. But if we don't intend for this to be fully integration tested then I'm not opposed to leaving it as is.

Reply:
The meaningful integration testing here would be between the pipeline and the databases, and the pipeline outputs and the segmentation app. I don't think we'll get a lot of benefit from unit-testing `run` alone.

Reply:
re: Isaak and cleaner test code. To test an ArgschemaParser class called `Something` like this you can simply do:
```
def test_something():
    args = {"a": 1, "b": 2}
    mysomething = Something(input_data=args, args=[])
    mysomething.run()
    assert(some_stuff)
```
I don't see how it is unclean. But, making the `run()` command just pass things to a function as function args eliminates the whole point of inheriting this class from ArgSchemaParser, and that was my design choice to make, IMO.

Reply:
I think this is to taste but to me the use of argschema is to allow ease of use in the command line while maintaining usability in imports. I don't like creating objects if they aren't explicitly needed and I think that running this from another module shouldn't need to use argschema if it's not necessary. It's to taste to some extent but if that's your design choice that's fine, I just prefer as much functional style over OOP as possible.